common/src/main/java/com/microsoft/identity/common/internal/eststelemetry/EstsTelemetry.java

```diff
@@ -194,122 +215,143 @@ private IRequestTelemetryCache createLastRequestTelemetryCache(@NonNull final Co
-     *
-     * @param correlationId      correlation id of the request
-     * @param acquireTokenResult the result obtained from the acquire token call
-     */
-    public void flush(final String correlationId, final AcquireTokenResult acquireTokenResult) {
-        final String errorCode = TelemetryUtils.errorFromAcquireTokenResult(acquireTokenResult);
-        flush(correlationId, errorCode);
-    }
+        final String correlationId = command.getParameters().getCorrelationId();
 
-    /**
-     * Flush the telemetry data associated to the correlation id to the {@link android.content.SharedPreferences} using the {@link SharedPreferencesLastRequestTelemetryCache}.
-     * Removes the telemetry associated to the correlation id from the telemetry map,
-     * and saves it to the cache (SharedPreferences) as the last request telemetry.
-     *
-     * @param correlationId correlation id of the request
-     * @param errorCode     error that may have occurred during the request
-     */
-    public void flush(final String correlationId, final String errorCode) {
-        final String methodName = ":flush";
         if (mTelemetryMap == null || correlationId == null) {
             return;
         }
 
-        RequestTelemetry currentTelemetry = mTelemetryMap.get(correlationId);
+        CurrentRequestTelemetry currentTelemetry = mTelemetryMap.get(correlationId);
+
         if (currentTelemetry == null) {
             return;
         }
 
-        RequestTelemetry lastTelemetry = setupLastFromCurrent(currentTelemetry);
-        lastTelemetry.putTelemetry(Schema.Key.CORRELATION_ID, correlationId);
-        lastTelemetry.putTelemetry(Schema.Key.ERROR_CODE, errorCode);
+        // load the last request object from cache
+        LastRequestTelemetry lastRequestTelemetry = loadLastRequestTelemetryFromCache();
+
+        // We did not have a last request object in cache, let's create a new one and derive
+        // fields from current request where applicable
+        if (lastRequestTelemetry == null) {
+            lastRequestTelemetry = new LastRequestTelemetry(currentTelemetry.getSchemaVersion());
+            lastRequestTelemetry = (LastRequestTelemetry) lastRequestTelemetry.derive(currentTelemetry);
+        }
+
+        if (isTelemetryLoggedByServer(command, commandResult)) {
+            // telemetry headers have been sent to token endpoint and logger by sts
+            // this is the time to reset local telemetry state
+
+            // reset silent successful count as we just went to token endpoint
+            lastRequestTelemetry.resetSilentSuccessCount();
+
+            // get the index for the last element of the failed request/error array so we can wipe
+            // all array elements up to this point (as that data is already in PerRequestTable)
+            // and doesn't need to be held onto or sent again
+            final int loggedIndex = getLoggedTelemetryIndex();
```

Comment:
I think it might be risky using index as we're supporting concurrent requests.

What if we have the command store a list of last request correlation ID, and if sent, then we remove the entry with the associated correlation ID from the cache?

Reply:
It is basically separation of concerns. I don't fully understand what you're suggesting but right now the data is stored in the SharedPreferences and persists. Are you suggesting to store in the Command? Commands are temporary so once those are over we would lose information.

Reply:
Let's say if we have more than 1 command running at the same time, and then they get the same value from getLoggedTelemetryIndex() and then try to wipe at the same time, would index out of bound occur?

I was suggesting storing the 'key' of failed request/error cache entry inside command, and then use those keys to wipe the values out. This would make it thread-safe. (the first wipe will win, the later wipes will not do anything).

Reply:
or there might be another way to address this. My concern is primarily the OutOfBound exception.

Reply:
OutOfBound should not occur as `wipe` is an instance method so even though these are concurrent requests, they each should have a different instance of the Last Request Telemetry object and hence when wiped is called, then it is called on the instance that is being used during the execution of that flush method. Inside wipe, we have a check like this for index:

`        if (index < 0 || index > failedRequests.size() || index > errors.size()) {
            return;
        }`

I think this should ensure that we're safe. What do you think?

Reply:
ahh.. ok. so you wipe from each instance then proceed to save those instances to the cache, and in this case the last one wins.

Ok in this case we should be safe from OOB. I'm still in the middle of PR and thought that you saved those request/error entries separately.

Reply:
The potential error that could happen in this case is a missing telemetry (i.e. There are 5 concurrent requests, the 3rd request failed, but the 5th overwrites the lastRequestTelemetry object).

Reply:
Having each command keeping track of what 'last request/error' it sent might be more straightforward IMHO. In this case, we can be certain we only wipe the ones that are already sent.

Reply:
@rpdome I've taken your suggestion and refactored the code. I didn't do at the command though as I believe telemetry should be kept separate from commands as much as possible due to separation of concerns. Also the currently executing command isn't available from everywhere. We would have to put the entire command object on the `DiagnosticContext` if we are to achieve that. I want to avoid doing that. So to solve this problem I've create a new `Set` at the telemetry layer that keeps track of all the failed requests sent to the server for the currently executing request and then later when we flush, we just remove all of these from the cache as these are sent. Two `FailedRequest` objects are considered the same if they have the same `correlationId` so I'm using that as you suggested to keep track of this. The index is no longer used and I've removed that. Thanks for the suggestion!